lib/srv/discovery/fetchers/kube_services.go

```diff
@@ -0,0 +1,335 @@
+			ctx,
+			metav1.ListOptions{
+				Continue: nextToken,
+			})
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		for _, s := range kubeServices.Items {
+			if !namespaceFilter(s.GetNamespace()) {
+				// Namespace is not in the list of namespaces to fetch
+				continue
+			}
+			match, _, err := services.MatchLabels(f.FilterLabels, s.Labels)
+			if err != nil {
+				return nil, trace.Wrap(err)
+			} else if match {
+				result = append(result, s)
+			} else {
+				f.Log.WithField("service_name", s.Name).Debug("Service doesn't match labels.")
+			}
+		}
+		nextToken = kubeServices.Continue
+		if nextToken == "" {
+			break
+		}
+	}
+	return result, nil
+}
+
+const (
+	protoHTTPS = "https"
+	protoHTTP  = "http"
+	protoTCP   = "tcp"
+)
+
+// Get fetches Kubernetes apps from the cluster
+func (f *KubeAppFetcher) Get(ctx context.Context) (types.ResourcesWithLabels, error) {
+	kubeServices, err := f.getServices(ctx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// Converting service to apps can involve performing a HTTP ping to the service ports to determine protocol.
+	// Both services and ports inside services are processed in parallel to minimize time.
+	// We also set limit to prevent potential spike load on a cluster in case there are a lot of services.
+	g, _ := errgroup.WithContext(ctx)
+	g.SetLimit(10)
+
+	// Convert services to resources
+	var (
+		appsMu sync.Mutex
+		apps   types.Apps
+	)
+	for _, service := range kubeServices {
+		service := service
+
+		// Skip kubernetes own internal services
+		if isInternalKubeService(service) {
+			continue
+		}
+
+		// Skip service if it has type annotation and it's not 'app'
+		if v, ok := service.GetAnnotations()[types.DiscoveryTypeLabel]; ok && v != services.KubernetesMatchersApp {
+			continue
+		}
+
+		g.Go(func() error {
+			protocolAnnotation := service.GetAnnotations()[types.DiscoveryProtocolLabel]
+
+			ports, err := getServicePorts(service)
+			if err != nil {
+				f.Log.WithError(err).Errorf("could not get ports for the service %q", service.GetName())
+				return nil
+			}
+
+			portProtocols := map[v1.ServicePort]string{}
+			for _, port := range ports {
+				protocol := ""
+				switch protocolAnnotation {
+				case protoHTTPS, protoHTTP, protoTCP:
+					protocol = protocolAnnotation
+				default:
+					protocol = autoProtocolDetection(getServiceFQDN(service), port, f.ProtocolChecker)
+					if protocol == protoTCP {
+						continue
+					}
+				}
+				portProtocols[port] = protocol
+			}
+
+			for port, portProtocol := range portProtocols {
+				if len(portProtocols) == 1 {
+					port.Name = ""
+				}
+				newApp, err := services.NewApplicationFromKubeService(service, f.ClusterName, portProtocol, port)
+				if err != nil {
+					f.Log.WithError(err).Warnf("Could not get app from a Kubernetes service %q, port %d", service.GetName(), port.Port)
+					return nil
+				}
+
+				appsMu.Lock()
```

Comment:
Appends the apps to a local array and lock only once you append all of them to global apps.

This helps the lock contention